add new plugin (rest-client) (#3)
Green-Software-Foundation/community-plugins: docs/rest-client.md

# REST client

`rest-client` is a plugin for executing HTTP methods `GET`, `PUT`, and `POST`. 

> You provide some query parameters including `url`, `method`, and so on. `url` is URL for an online resource. `method` specifies the HTTP method. SSL/TLS is turned off in this plugin.

Author: I fixed it. Now you can choose using TLS or not.

> When `GET` is used, only a numeric value is accepted as `output`, otherwise an error is returned. When `PUT` and `POST` are used, the `output` is the HTTP response.

Author: I changed HTTP response to response data as output.

## Parameters

### Plugin config

> The parameters are required in config: `method`, `url`, `data`, `bearer-tokken`, `http-basic-authentication`, `jpath` and `output`.

Reviewer: Typo
```suggestion
The parameters are required in config: `method`, `url`, `data`, `bearer-token`, `http-basic-authentication`, `jpath` and `output`.
```

- `method`: HTTP method, you can choose `GET`, `PUT`, and `POST`
- `url`: The URL to which you send the request 
- `data` : data to be sent to the server, used with `PUT` and `POST` methods

> - `bearer-tokken`: optional and Bearer Token providing authentication information to the target.

Reviewer: ```suggestion
- `bearer-token`: optional and Bearer Token providing authentication information to the target.
```

  - `url`: The URL to which you send the request
  - `headers`: optional and HTTP request header
    - `Authorization`: optional and provide authentication information to the target. For example,
      - Bearer Token Authentication: 
        ```yml
        Authorization: Bearer <token>
        ```
      - API Key Authentication: 
        ```yml
        Authorization: ApiKey <api_key>
        ```
  - `data` : request body, used with `PUT` and `POST` methods
  - `auth` : optional and authentication information for HTTP Basic Authentication
    ```yml
    auth: {
      username: 'yourUsername',
      password: 'yourPassword'
    }
    ```
- `jpath`: optional, JSONPath expression, and you can use it when using the GET method
- `output`: parameter name to store the result of GET method in the output array and you only need to configure it when using the GET method

### Inputs

There are no strict requirements on input for this plugin because they depend upon the contents of the target and your input data at the time the rest client is invoked. Please make sure you are requesting data from target that exist in the target database.

## Returns

> input data with value requested by `GET` method or input data for `PUT` and `POST`

Author: You're right. I fixed the sentence to ''`output`: numerical value requested by `GET` and scraped by `jpath` or HTTP response for `PUT` and `POST`''

## Example manifest

IF users will typically call the plugin as part of a pipeline defined in a manifest file. In this case, instantiating the plugin is handled by `if-run` and does not have to be done explicitly by the user. The followings are example manifests that calls `rest-client`:

```yaml
name: get-wattge
description: successful path
tags:
initialize:
  plugins:
    get-wattage:
      path: 'https://github.com/Green-Software-Foundation/community-plugins'
      method: RESTClient
      config:
        query:
          url: https://api.example.com/data
          method: get
          headers: 
            Authorization: Bearer your-secret-token
        jpath: $.information[?(@.id==1)].wattage
        output: wattage
tree:
  children:
    child:
      pipeline:
        compute:
          - get-wattage
      inputs:
        - timestamp: 2023-07-06T00:00 
          duration: 100
```
```yaml

> name: put-wattge

Author: Thank you for your advice. I corrected it.

description: successful path
tags:
initialize:
  plugins:
    put-wattage:
      path: 'https://github.com/Green-Software-Foundation/community-plugins'
      method: RESTClient
      config:
        query:
          method: put
          url: https://api.example.com/data

>           data: {wattage: 100}

Author: `data` is data to be sent to the server. So your understanding is correct. I fixed the explanation of `data`. I modified the code to make it possible to get the HTTP response from `https://api.example.com/data` as an output parameter.

          auth: {
            username: 'yourUsername',
            password: 'yourPassword'
          }
tree:
  children:
    child:
      pipeline:
        compute:
          - put-wattage
      inputs:
        - timestamp: 2023-07-06T00:00 
          duration: 100

```
```yaml

> name: post-wattge

Author: Thank you for your advice. I corrected it.

# REST client

`rest-client` is a plugin for executing HTTP methods `GET`, `PUT`, and `POST`. 

You provide some query parameters including `url`, `method`, and so on. `url` is URL for an online resource. `method` specifies the HTTP method. Only a numeric value is accepted as `output`, otherwise an error is returned.

## Parameters

### Plugin config

The parameters are required in config: `method`, `url`, `data`, `headers`, `http-basic-authentication`, `tls-verify`, `jpath` and `output`.

- `method`: HTTP method, you can choose `GET`, `PUT`, and `POST`
- `url`: the URL to which you send the request 

> - `data` : used with `PUT` and `POST` methods and the data specified in `data` is set in the request body in JSON format but if `Content-Type` is specified in `headers`, it can be overridden by that value.

Reviewer: I guess you implies `application/json` would be set in `Content-Type` if `PUT` or `POST` are specified in `method` by default, right?

Author: Sorry, I forgot to add this function. I add it.

- `headers`: optional and request header
    ```yml
    headers: 
      - Accept: application/json
      - Authorization: Bearer xxxxxx
      - X-Custom-Header: foo
    
    ```
- `http-basic-authentication` : optional and authentication information for HTTP Basic Authentication
    ```yml
    http-basic-authentication: 
      username: 'yourUsername'
      password: 'yourPassword'
    
    ```

> - `ssl-tls-active`: optional, SSL/TLS is enabled if true and undefined, SSL/TLS is disabled if false

Reviewer: Again, `rejectUnauthorized` in `https` module does not mean "disabling SSL/TLS", it means skip certificates verification. https://nodejs.org/docs/latest/api/tls.html

Author: Thank you. I understand the meaning. I fixed it.

> - `tls-verify`: optional, the server certificate is verified against the list of supplied CAs if true and undefined

Reviewer: ```suggestion
- `tls-verify`: optional, the server certificate is verified against the list of supplied CAs if true or undefined. `true` is set by default.
```

> - `ssl-tls-active`: optional, SSL/TLS certificate is validated if true and undefined and SSL/TLS certificate validation is skiped  if false

Reviewer: As I said before, this parameter affects whether verify certifications because it is used for `rejectUnauthorized`. Thus I think it is better to rename to `tls-verify` and rewrite description to represent this behavior. You can refer [Node.js document](https://nodejs.org/docs/latest/api/tls.html).

Author: I understand it it related to that the server certificate is verified against the list of supplied CAs if true. I fixed name and description.
